pkg/pattern/tokenization/replacer.go

```diff
@@ -0,0 +1,859 @@
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// ANSIC, UnixDatem, RubyDate e.g
+		// - "Mon Jan  2 15:04:05 2006"
+		// - "Mon Jan  2 15:04:05 MST 2006"
+		// - "Mon Jan 02 15:04:05 -0700 2006"
+		if r.advanceChar(' ') && r.advanceMonthName() && r.advanceChar(' ') && (r.advanceChar(' ') || true) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(true) && r.advanceChar(' ') && r.advanceYear() {
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// Linux, e.g.
+		// - "Mon  2 Jan 15:04:05 MST 2006"
+		// - "Tue 23 Jan 15:04:05 -0700 2023"
+		if r.advanceChar(' ') && (r.advanceChar(' ') || true) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceMonthName() && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(false) && r.advanceChar(' ') && r.advanceYear() {
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.resetHead()
+
+		// RFC850, e.g.
+		//  - "Monday, 02-Jan-06 15:04:05 MST"
+		backtrackedSlice := r.source[r.head-3:]
+		var matchedDay []byte
+		for _, dw := range longDayNames {
+			if bytes.HasPrefix(backtrackedSlice, dw) {
+				matchedDay = dw
+				break
+			}
+		}
+		if matchedDay != nil {
+			r.head += len(matchedDay) - 3
+			if r.advanceChar(',') && r.advanceChar(' ') && r.advanceUintUpTo(31, 2) && r.advanceChar('-') && r.advanceMonthName() && r.advanceChar('-') && r.advanceUintUpTo(99, 2) && r.advanceChar(' ') && r.advanceTime(false) && r.advanceStringOrNumericTimeZone(true) {
+				r.emit(false, placeholderTimestamp)
+				return true
+			}
+		}
+
+		r.cur -= 3 // unconsume
+		r.resetHead()
+		return false
+	}
+	r.resetHead()
+
+	if r.advanceMonthName() {
+		r.commit() // provisionally consume this
+
+		// Linux journald logs and others similar like this:
+		//  - Feb 29 23:00:14
+		//  - Apr-10 23:43:46.807
+		//  - Jul  1 00:21:28
+		if (r.advanceChar('-') || (r.advanceChar(' ') && (r.advanceChar(' ') || true))) && r.advanceUintUpTo(31, 2) && r.advanceChar(' ') && r.advanceTime(false) {
+			r.commit()
+			// This is already a timestamp, but let's try to match subseconds as well
+			if r.advanceChar('.') && r.advanceUint() {
+				r.commit()
+			} else {
+				r.resetHead()
+			}
+			r.emit(false, placeholderTimestamp)
+			return true
+		}
+		r.cur -= 3 // unconsume
+		r.resetHead()
+		return false
+	}
+	r.resetHead()
+
+	return false
+}
+
+func (r *replacer) wasNumPreservingKeyword() bool {
+	for _, key := range numPreservingKeys {
+		pPos := r.cur - 1 - len(key)
+		if pPos < -1 {
+			return false // all subsequent keys are longer
+		}
+		if pPos != -1 && !boundaryChars[r.source[pPos]] {
+			continue
+		}
+		if bytes.HasPrefix(r.source[pPos+1:], key) {
```

Comment:
This probably can be optimized, the `numPreservingKeys` values can be of a `string` type and if this check is something like `string(r.source[pPos+1:pPos+len(key)]) == key)`, it might be faster. This should not result in an actual string allocation because of [some Go compiler optimizations](https://github.com/golang/go/blob/5419f652b6593a0916f7b2155435b945e8ee0fb4/src/runtime/string.go#L136-L150) for cases like this. Besides, this is basically what `bytes.HasPrefix` does internally as well, just with an extra bounds check and function calls... :sweat_smile:  Which are probably inlined, but it can't hurt to compare the values directly :man_shrugging: